### ISSUE TITLE
Setup automated build environment/CI

### ISSUE THREAD

**Issue by julianrichen** [winepak/winepak#3]
Possible options:
* [Buildbot](https://buildbot.net/)
* [Travis CI](https://travis-ci.org/)
* [Circle CI](http://circleci.com/)
* [Jenkins](https://jenkins.io/)

**Comment by julianrichen** [comment]
WIP Travis CI with winepak/winepak-sdk-images@43dffd11243fbeffce3bc6a90db267fe5d281195

Currently building:
https://travis-ci.org/winepak/winepak-sdk-images/builds/383012616

Notice: I have no idea if the `.travis.yml` is right and I have never worked with docker before so I guess we'll see how far it builds...

---
Also @paulcarroty is trying Circle CI:
https://github.com/paulcarroty/winepak-sdk-images/blob/master/.circleci/config.yml

**Comment by paulcarroty** [comment]
TravisCI free account is limited for 40m, not enough for build wine&modules from source.

Actually, your [howto](https://github.com/winepak/winepak-sdk-images/blob/master/README.md) doesn't works inside docker.

**Comment by julianrichen** [comment]
Ah. When comparing the two services Travis CI didn't mention anything about build timeouts on the front page and they seemed to indicate you can run unlimited builds, however, you're put into a queue. I now [see they list the build timeouts in the docs](https://docs.travis-ci.com/user/customizing-the-build/#Build-Timeouts). At the time it seemed better than the x amount of build minute per month that circle ci offers.

That being the case I'll look into circle. 1,500 build minutes per month is still very generous.

**Comment by paulcarroty** [comment]
CircleCI IMHO is the best today. In critical circumstances you can use different CI for each branch.

**Comment by julianrichen** [comment]
Yah I think I may use Travis CI for applications since they take seconds to build but we may have a lot of request. Circle CI is really nice so far, very clean UI and config file.

Working on build: https://circleci.com/gh/winepak/winepak-sdk-images/5

```
Committing stage init to cache
Starting build of org.winepak.Sdk
fuse: device not found, try 'modprobe fuse' first
========================================================================
Building module platform-setup in /root/repo/.flatpak-builder/build/platform-setup-1
========================================================================
]2;flatpak-builder: Building platform-setup]2;flatpak-builder: Installing platform-setupRunning: mkdir -p /usr/lib/32bit
error: Build directory /root/repo/.flatpak-builder/rofiles/rofiles-cB9kYK not initialized, use flatpak build-init
Error: module platform-setup: Child process exited with code 1
Exited with code 1
```

Same as you. I think `flatpak-builder` needs `--disable-rofiles-fuse`. Looking online fuse support seems to vary in docker?

Re-built with option: https://circleci.com/gh/winepak/winepak-sdk-images/6

```
Committing stage init to cache
Starting build of org.winepak.Sdk
========================================================================
Building module platform-setup in /root/repo/.flatpak-builder/build/platform-setup-1
========================================================================
]2;flatpak-builder: Building platform-setup]2;flatpak-builder: Installing platform-setupRunning: mkdir -p /usr/lib/32bit
bwrap: Creating new namespace failed: Operation not permitted
Error: module platform-setup: Child process exited with code 1
Exited with code 1
```
[This issue suggest](https://github.com/flatpak/flatpak/issues/1326) you then need to add `--privileged` when running docker, so:

1. Not sure how to pass options to docker run in circle ci
2. flatpak should be able to build without `--privileged`

Will look into it later

----

Also need to look into caching the download & install phase for fd.o and modules.

**Comment by paulcarroty** [comment]
> Not sure how to pass options to docker run in circle ci

Use `machine`, then run docker manually in privileged mode. Just copy from Travis config.

**Comment by paulcarroty** [comment]
https://circleci.com/docs/2.0/status-badges/

**Comment by paulcarroty** [comment]
@julianrichen [Fixed](https://github.com/paulcarroty/winepak-sdk-images/blob/master/.circleci/config.yml).

**Comment by paulcarroty** [comment]
SDK 3.0 i386 takes 50m for building what beats 2h limit. As I said before, building wine is absolute madness in this case.

**Comment by julianrichen** [comment]
> @julianrichen [Fixed](https://github.com/paulcarroty/winepak-sdk-images/blob/master/.circleci/config.yml).

Looks good, create a PR and I'll merge it.

> SDK 3.0 i386 takes 50m for building what beats 2h limit. As I said before, building wine is absolute madness in this case.

Wine is large and Platform/Sdks also take time to "commit", it's to be expected. We won't be building them often and the total should only be 150 ~ 200 minutes (50 + 50 + 40 + 40). The wine extensions don't take nearly as long and the Compat32 extensions are just copying the i386 branch to a x86_64 branch.

Also what do you mean by "what beats 2h limit"? CircleCI has a 2h build limit? I can't find reference to that on the FAQ. Also on the main page it says:

> We offer a total of four free linux containers ($2400 annual value) for open-source projects. Simply keeping your project public will enable this for you!

Might be able to split the four builds into four containers for parallel building.

**Comment by paulcarroty** [comment]
2h is hardcoded limit, you can change it. Probably.
https://discuss.circleci.com/t/timeout-more-than-120-minutes/640

**Comment by julianrichen** [comment]
> Test suites are conventionally defined at the job level in your .circleci/config.yml file. The parallelism key specifies how many independent executors will be set up to run the steps of a job.
> 
> To run a job’s steps in parallel, set the parallelism key to a value greater than 1.

[Source](https://circleci.com/docs/2.0/parallelism-faster-jobs/)

So just need to make the Compat32 steps wait until i386 builds are done.

Other things the config should have (if possible):
* Environment variable for the wine-staging version, [Wine 3.9](https://www.winehq.org/announce/3.9) was just released and being able to build multiple versions would be good.
  * + No reason to re-build an older version unless we change something critical
* Make some steps togglable
  * e.g. We don't need to build `org.winepak.(Sdk|Platform) every time.

**Comment by paulcarroty** [comment]
> We won't be building them often

Every 2 weeks plus some apps needs old versions.

Now I play with wine builds from PlayOnLinux, 'cause there's no reason to lost a lot of time. And switch to another version is easy and fast.

**Comment by julianrichen** [comment]
> Every 2 weeks plus some apps needs old versions.

For staging yes, that's an i386 and x86_64 build every 2 weeks 80 ~ 100 minutes. But the old versions won't be changing. No point in re-building something that won't be changed.

> Now I play with wine builds from PlayOnLinux, 'cause there's no reason to lost a lot of time. And switch to another version is easy and fast.

You can do that but the versions that will be released in the winepak repo will be from source. As mentioned before I'm not using pre-built binaries.

**Comment by Calinou** [comment]
What about [GitLab CI](https://about.gitlab.com/features/gitlab-ci-cd/)? The default timeout is set to 60 minutes, ~~but it can be set to any value (I set it to 180 minutes in one of my projects).~~ It also allows running an unlimited amount of pipelines at once (up to 50,000 minutes per month).

As of recently, it can use a GitHub repository as a source so you don't need to configure a mirror repository on GitLab anymore.

**Update (2018-08-14):** GitLab CI's shared runners are now limited to 60 minutes per job. :cry:

**Comment by MihaiTabara** [comment]
@julianrichen Hi! Out of curiosity, did you find a way to overcome the `bwrap permissions issues` mentioned [here](https://github.com/winepak/winepak/issues/3#issuecomment-391793044) without enabling the container access to all devices on the host? In my current setting, I don't have the option to send arguments to `docker run`. 

Thanks!

**Comment by johnramsden** [comment]
If you're interested, I can let you know how I approached this using TravisCI. I had originally planned on writing a blog post as a follow-up to one I had written about [packaging Path of Exile](https://ramsdenj.com/2018/03/26/packaging-pathofexile-with-flatpak.html), but I never got around to finishing it. Here's my [.travis.yml](https://github.com/johnramsden/flatpakrepo.johnramsden.ca/blob/master/.travis.yml) configuration if it helps.

In simple terms I had a [flatpak repo](https://github.com/johnramsden/flatpakrepo) that was a git sub-module. When I push to my [flatpakrepo](https://github.com/johnramsden/flatpakrepo.johnramsden.ca), Travis CI would automatically pull from each sub-module and [build each flatpak](https://github.com/johnramsden/flatpakrepo.johnramsden.ca/blob/master/scripts/build.sh).

**Comment by MihaiTabara** [comment]
Thanks for following-up my comment, I was out a few days. I managed to get this done by running the docker container with `--priviledge` mode on. Thanks for the explanations!
